accounts-db/src/append_vec.rs

```diff
@@ -552,6 +561,47 @@ impl AppendVec {
         self.path.clone()
     }
 
+    /// help with the math of offsets when navigating the on-disk layout in an AppendVec.
+    /// data is at the end of each account and is variable sized
+    /// the next account is then aligned on a 64 bit boundary.
+    /// With these helpers, we can skip over reading some of the data depending on what the caller wants.
+    fn next_account_offset(start_offset: usize, stored_meta: &StoredMeta) -> AccountOffsets {
+        let start_of_data = start_offset
+            + std::mem::size_of::<StoredMeta>()
+            + std::mem::size_of::<AccountMeta>()
+            + std::mem::size_of::<AccountHash>();
+        let aligned_data_len = u64_align!(stored_meta.data_len as usize);
+        let next_account_offset = start_of_data + aligned_data_len;
+        let offset_to_end_of_data = start_of_data + stored_meta.data_len as usize;
+
+        AccountOffsets {
+            next_account_offset,
+            offset_to_end_of_data,
+        }
+    }
+
+    /// iterate over all pubkeys and call `callback`.
+    /// This iteration does not deserialize and populate each field in `StoredAccountMeta`.
+    /// `data` is completely ignored, for example.
+    /// Also, no references have to be maintained/returned from an iterator function.
+    /// This fn can operate on a batch of data at once.
+    pub(crate) fn scan_pubkeys(&self, mut callback: impl FnMut(&Pubkey)) {
```

Comment:
Can we change the callback to take Pubkey instead of a &Pubkey? 

```
-    pub(crate) fn scan_pubkeys(&self, mut callback: impl FnMut(&Pubkey)) {
+    pub(crate) fn scan_pubkeys(&self, mut callback: impl FnMut(Pubkey)) {
```

`store_meta` already holds a copy of the pubkey. Instead of passing a ref to store_meta.pubkey to the callback, we can `move` the pubkey into the callback. This will save the extra copy of pubkey inside the callback later, for example. 

```
                         store.accounts.scan_pubkeys(|k| {
-                            pubkeys.insert(*k);
+                            pubkeys.insert(k);
                         });
```

Reply:
The whole diff. 
```
diff --git a/accounts-db/src/accounts_db.rs b/accounts-db/src/accounts_db.rs
index beabcb72f5..53772fb433 100644
--- a/accounts-db/src/accounts_db.rs
+++ b/accounts-db/src/accounts_db.rs
@@ -2903,7 +2903,7 @@ impl AccountsDb {
                         }
                         oldest_dirty_slot = oldest_dirty_slot.min(*slot);
                         store.accounts.scan_pubkeys(|k| {
-                            pubkeys.insert(*k);
+                            pubkeys.insert(k);
                         });
                     });
                     oldest_dirty_slot
@@ -16671,7 +16671,7 @@ pub mod tests {
                 // Note that we assume traversals are both in the same order, but this doesn't have to be true.
                 let mut compare = Vec::default();
                 storage.accounts.scan_pubkeys(|k| {
-                    compare.push(*k);
+                    compare.push(k);
                 });
                 assert_eq!(compare, vec.iter().map(|(k, _)| *k).collect::<Vec<_>>());
                 vec
diff --git a/accounts-db/src/accounts_file.rs b/accounts-db/src/accounts_file.rs
index 0cd11296ce..5d1d25ad34 100644
--- a/accounts-db/src/accounts_file.rs
+++ b/accounts-db/src/accounts_file.rs
@@ -157,7 +157,7 @@ impl AccountsFile {
     }
 
     /// iterate over all pubkeys
-    pub(crate) fn scan_pubkeys(&self, callback: impl FnMut(&Pubkey)) {
+    pub(crate) fn scan_pubkeys(&self, callback: impl Fn(Pubkey)) {
         match self {
             Self::AppendVec(av) => av.scan_pubkeys(callback),
             Self::TieredStorage(_) => unimplemented!(),
diff --git a/accounts-db/src/append_vec.rs b/accounts-db/src/append_vec.rs
index 5e26bf1849..8b2abd6e1b 100644
--- a/accounts-db/src/append_vec.rs
+++ b/accounts-db/src/append_vec.rs
@@ -585,7 +585,7 @@ impl AppendVec {
     /// `data` is completely ignored, for example.
     /// Also, no references have to be maintained/returned from an iterator function.
     /// This fn can operate on a batch of data at once.
-    pub(crate) fn scan_pubkeys(&self, mut callback: impl FnMut(&Pubkey)) {
+    pub(crate) fn scan_pubkeys(&self, callback: impl Fn(Pubkey)) {
         let mut offset = 0;
         loop {
             let Some((stored_meta, _)) = self.get_type::<StoredMeta>(offset) else {
@@ -597,7 +597,7 @@ impl AppendVec {
                 // data doesn't fit, so don't include this pubkey
                 break;
             }
-            callback(&stored_meta.pubkey);
+            callback(stored_meta.pubkey);
             offset = next.next_account_offset;
         }
     }
```

Reply:
will this perf be better somehow?

Reply:
I think so. 
`mut callback: impl FnMut(&Pubkey)`, the pubkey is captured by &mut. 
When ` compare.push(*k);` is called, it can't move `k` because of the &mut. Instead, a copy will have to be made. Pubkey impelments copy that's why it compile. If Pubkey doesn't implement copy, the compile will fail. 

The following link is an example. 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=441e5c710916629e79dd6820dc9dae89

Foo didn't drive copy, then compile fails to move.

Reply:
i know it copies in this particular fnmut. not all fnmuts may choose to copy. the source entry is &struct where the struct has a pubkey. so, since the whole struct is a ref, the struct doesn't own a copy of the pubkey. the pubkey will be copied no matter what when we assign it into compare.push. I think the key part is that the source of the pubkey in StoredMeta or whatever is & to a local, temporary file buf.

Reply:
Ah. I see. You are right. Thanks for the explaination.